File: src/main/scala/geotrellis/RasterExtent.scala
package geotrellis

import scala.math.{min, max, round, ceil, floor}

case class GeoAttrsError(msg:String) extends Exception(msg)

/**
 * Represents grid coordinates of a subsection of a RasterExtent.
 * These coordinates are inclusive.
 */
case class GridBounds(colMin:Int,rowMin:Int,colMax:Int,rowMax:Int)

/**
 * RasterExtent objects represent the geographic extent (envelope) of a raster.
 *
 * The Raster extent has two coordinate concepts involved: map coordinates and grid
 * coordinates. Map coordinates are what the [[Extent]] class uses, and specifies points
 * using an X coordinate and a Y coordinate. The X coordinate is oriented along west to east
 * such that the larger the X coordinate, the more eastern the point. The Y coordinate is
 * along south to north such that the larger the Y coordinate, the more Northern the point.
 *
 * This contrasts with the grid coordinate system. The grid coordinate system does not
 * actually reference points on the map, but instead a cell of the raster that represents
 * values for some square area of the map. The column axis is similar in that the number
 * gets larger as one goes from west to east; however, the row axis is inverted from map coordinates:
 * as the row number increases, the cell is heading south. The top row is labeled as 0, and the next
 * 1, so that the highest indexed row is the southern most row of the raster.
 * A cell has a height and a width that is in terms of map units. You can think of it as each cell
 * is itself an extent, with width cellwidth and height cellheight. When a cell needs
 * to be represented or thought of as a point, the center of the cell will be used.
 * So when gridToMap is called, what is returned is the center point, in map coordinates.
 *
 * Map points are considered to be 'inside' the cell based on these rules:
 *  - If the point is inside the area of the cell, it is included in the cell.
 *  - If the point lies on the north or west border of the cell, it is included in the cell.
 *  - If the point lies on the south or east border of the cell, it is not included in the cell,
 *    it is included in the next southern or eastern cell, respectively.
 *
 * Note that based on these rules, the eastern and southern borders of an Extent are not actually
 * considered to be part of the RasterExtent.
 */
case class RasterExtent(extent:Extent, cellwidth:Double, cellheight:Double, cols:Int, rows:Int) {

  if (cellwidth  <= 0.0) throw GeoAttrsError("invalid cell-width")
  if (cellheight <= 0.0) throw GeoAttrsError("invalid cell-height")
  if (cols <= 0) throw GeoAttrsError("invalid cols")
  if (rows <= 0) throw GeoAttrsError("invalid rows")

  val width  = extent.width
  val height = extent.height

  /**
   * The size of the extent, e.g. cols * rows.
   */
  def size = cols * rows

  /**
   * Compare this object with another GeoAttrs object, as per the comparison
   * rules in Extent#compare.
   */
  def compare(other:RasterExtent) = extent.compare(other.extent)

  /**
   * Determine if the underlying extent contains the given point.
   */
  def containsPoint(x:Double, y:Double) = extent.containsPoint(x, y)
  
  /**
   * Convert map coordinates (x,y) to grid coordinates (col,row).
   */
  def mapToGrid(x:Double, y:Double) = {
    val col = ((x - extent.xmin) / cellwidth).toInt
    val row = ((extent.ymax - y) / cellheight).toInt
    (col, row)
  }

  /**
   * Convert map coordinate x to grid coordinate column.
   */
  def mapXToGrid(x:Double) = mapXToGridDouble(x).toInt
  
  def mapXToGridDouble(x:Double) = (x - extent.xmin) / cellwidth

    
  /**
   * Convert map coordinate y to grid coordinate row.
   */
  def mapYToGrid(y:Double) = mapYToGridDouble(y).toInt
  
  /**
   * 
   */
  def mapYToGridDouble(y:Double) = (extent.ymax - y ) / cellheight
  
  /**
   * Convert map coordinate tuple (x,y) to grid coordinates (col,row).
   */
  def mapToGrid(mapCoord:(Double,Double)):(Int,Int) = {
    val (x,y) = mapCoord;
    mapToGrid(x,y)
  }

  // TODO: try to remove calls to max and min if possible
  /**
    * The map coordinate of a grid cell is the center point.
    */  
  def gridToMap(col:Int, row:Int) = {
    val x = max(min(col * cellwidth + extent.xmin + (cellwidth / 2), extent.xmax), extent.xmin)
    val y = min(max(extent.ymax - (row * cellheight) - (cellheight / 2), extent.ymin), extent.ymax)
    (x, y)
  }

  /**
   * Gets the GridBounds for this RasterExtent that is the smallest subgrid
   * containing all points within the extent. The extent is considered inclusive
   * on it's north and west borders, exclusive on it's east and south borders.
   * See [[RasterExtent]] for a discussion of grid and extent boundary concepts.
   */
  def gridBoundsFor(subExtent:Extent):GridBounds = {
    if(!extent.containsExtent(subExtent)) { throw ExtentRangeError("") }
    // West and North boundarys are a simple mapToGrid call.
    val (colMin,rowMin) = mapToGrid(subExtent.xmin, subExtent.ymax)

    // If South East corner is on grid border lines, we want to still only include
    // what is to the West and\or North of the point. However if the border point
    // is not directly on a grid division, include the whole row and/or column that
    // contains the point.
    val colMax = ceil((subExtent.xmax - extent.xmin) / cellwidth).toInt - 1
    val rowMax = ceil((extent.ymax - subExtent.ymin) / cellheight).toInt - 1
    
    GridBounds(max(0,colMin),
               max(0,rowMin),
               min(cols-1,colMax),
               min(rows-1,rowMax))
  }
  
  /**
   * Combine two different GeoAttrs (which must have the same cellsizes).
   * The result is a new extent at the same resolution.
   *
   * TODO: this version currently warps the grid. we need two versions.
   *
   * TODO: relatedly, the translate version should require the grids to be
   * properly aligned.
   */
  def combine (that:RasterExtent) = {
    if (cellwidth != that.cellwidth)
      throw GeoAttrsError("illegal cellwidths: %s and %s".format(cellwidth, that.cellwidth))
    if (cellheight != that.cellheight)
      throw GeoAttrsError("illegal cellheights: %s and %s".format(cellheight, that.cellheight))

    val newExtent = extent.combine(that.extent)
    val newRows = ceil(newExtent.height / cellheight).toInt
    val newCols = ceil(newExtent.width / cellwidth).toInt

    RasterExtent(newExtent, cellwidth, cellheight, newCols, newCols)
  }
}

File: geotrellis-python/geotrellis/raster/RasterExtent.py
from __future__ import absolute_import
from geotrellis.spark.io.package_scala import ExceptionWithCause
from geotrellis.raster.GridDefinition import GridDefinition
from geotrellis.raster.Tile import Tile
from geotrellis.vector.Extent import Extent

import math

class GeoAttrsError(ExceptionWithCause):
    def __init__(self, msg, **kwargs):
        ExceptionWithCause.__init__(self, msg, **kwargs)

class RasterExtent(GridDefinition):
    def __init__(self, extent, cellwidth, cellheight, cols, rows):
        if cols < 0:
            raise GeoAttrsError("invalid cols: {0}".format(cols))
        if rows < 0:
            raise GeoAttrsError("invalid rows: {0}".format(rows))
        self.extent = extent
        self.cellwidth = cellwidth
        self.cellheight = cellheight
        self.cols = cols
        self.rows = rows

    def __eq__(self, other):
        if not isinstance(other, RasterExtent):
            return False
        return (self.extent == other.extent and
                self.cellwidth == other.cellwidth and
                self.cellheight == other.cellheight and
                self.cols == other.cols and
                self.rows == other.rows)

    def __hash__(self):
        return hash((self.extent, self.cellwidth, self.cellheight,
            self.cols, self.rows))

    @property
    def size(self):
        return self.cols * self.rows

    @property
    def dimensions(self):
        return (self.cols, self.rows)

    def mapToGrid(self, x, y = None):
        if y is None:
            mapCoord = x
            x, y = mapCoord
        col = int(math.floor((x - self.extent.xmin) / self.cellwidth))
        row = int(math.floor((self.extent.ymax - y) / self.cellheight))
        return (col, row)

    def mapXToGrid(self, x):
        return int(math.floor(self.mapXToGridDouble(x)))

    def mapXToGridDouble(self, x):
        return (x - self.extent.xmin) / self.cellwidth

    def mapYToGrid(self, y):
        return int(math.floor(self.mapYToGridDouble(y)))

    def mapYToGridDouble(self, y):
        return (self.extent.ymax) / self.cellheight

    def gridToMap(self, col, row):
        x = col * self.cellwidth + self.extent.xmin + (self.cellwidth / 2)
        y = self.extent.ymax - (row * self.cellheight) - (self.cellheight / 2)
        return (x, y)

    def gridColToMap(self, col):
        return col * self.cellwidth + self.extent.xmin + (self.cellwidth / 2)

    def gridRowToMap(self, row):
        return self.extent.ymax - (row * self.cellheight) - (self.cellheight / 2)

    def gridBoundsFor(self, subExtent, clamp = True):
        colMin, rowMin = self.mapToGrid(subExtent.xmin, subExtent.ymax)
        colMaxDouble = self.mapXToGridDouble(subExtent.xmax)
        if abs(colMaxDouble - math.floor(colMaxDouble)) < RasterExtent.epsilon:
            colMax = int(colMaxDouble) - 1
        else:
            colMax = int(colMaxDouble)

        rowMaxDouble = self.mapYToGridDouble(subExtent.ymin)
        if abs(rowMaxDouble - math.floor(rowMaxDouble)) < RasterExtent.epsilon:
            rowMax = int(rowMaxDouble) - 1
        else:
            rowMax = int(rowMaxDouble)

        if clamp:
            return GridBounds(
                    max(colMin, 0),
                    max(rowMin, 0),
                    min(colMax, cols-1),
                    min(rowMax, rows-1))
        else:
            return GridBounds(colMin, rowMin, colMax, rowMax)

    def combine(self, that):
        if self.cellwidth != that.cellwidth:
            raise GeoAttrsError("illegal cellwidths: {this} and {that}".format(
                this = self.cellwidth, that = that.cellwidth))
        if self.cellheight != that.cellheight:
            raise GeoAttrsError("illegal cellheights: {this} and {that}".format(
                this = self.cellheight, that = that.cellheight))

        newExtent = self.extent.combine(that.extent)
        newRows = int(math.ceil(newExtent.height / self.cellheight))
        newCols = int(math.ceil(newExtent.width / self.cellwidth))

        return RasterExtent(newExtent, self.cellwidth, self.cellheight, newCols, newRows)

    def withResolution(self, first, second = None):
        if second is None:
            cellSize = first
            targetCellWidth = cellSize.width
            targetCellHeight = cellSize.height
        else:
            targetCellWidth, targetCellHeight = first, second

        newCols = int(math.ceil((self.extent.xmax - self.extent.xmin) / targetCellWidth))
        newRows = int(math.ceil((self.extent.ymax - self.extent.ymin) / targetCellHeight))
        return RasterExtent(self.extent, targetCellWidth, targetCellHeight, newCols, newRows)

    def withDimensions(self, targetCols, targetRows):
        return RasterExtent.applyStatic(self.extent, targetCols, targetRows)

    def adjustTo(self, tileLayout):
        totalCols = tileLayout.tileCols * tileLayout.layoutCols
        totalRows = tileLayout.tileRows * tileLayout.layoutRows

        resampledExtent = Extent(self.extent.xmin,
                self.extent.ymax - (self.cellheight * totalRows),
                self.extent.xmin + (self.cellwidth * totalCols),
                self.extent.ymax)

        return RasterExtent(resampledExtent, self.cellwidth, self.cellheight, totalCols, totalRows)

    def rasterExtentFor(self, gridBounds):
        xminCenter, ymaxCenter = self.gridToMap(gridBounds.colMin, gridBounds.rowMin)
        xmaxCenter, yminCenter = self.gridToMap(gridBounds.colMax, gridBounds.rowMax)
        hcw, hch = self.cellwidth / 2, self.cellheight / 2
        e = Extent(xminCenter - hcw, yminCenter - hch, xmaxCenter + hcw, ymaxCenter + hch)
        return RasterExtent(e, self.cellwidth, self.cellheight, gridBounds.width, gridBounds.height)

    epsilon = 0.0000001

    @staticmethod
    def applyStatic(first, second, third = None):
        if third is None:
            if isinstance(first, Tile):
                tile = first
                extent = second
                return RasterExtent.applyStatic(extent, tile.cols, tile.rows)
            elif isinstance(second, Tile):
                extent = first
                tile = second
                return RasterExtent.applyStatic(extent, tile.cols, tile.rows)
            else:
                extent = first
                cellSize = second
                cols = int(extent.width / cellSize.width)
                rows = int(extent.height / cellSize.height)
                return RasterExtent(extent, cellSize.width, cellSize.height, cols, rows)
        else:
            extent = first
            cols, rows = second, third
            cw = extent.width / cols
            ch = extent.height / rows
            return RasterExtent(extent, cw, ch, cols, rows)
